### ISSUE TITLE
Fails to promote to staging

### ISSUE THREAD

**Issue by camusd** [jenkins-x/jx#493]
```
[web] Running shell script
++ cat ../../VERSION
+ jx promote -b --all-auto --timeout 1h --version 0.0.4
Promoting app tabletoprank-web version 0.0.4 to namespace jx-staging
Using pipeline: ****/tabletoprank-web/master build: #1
Cloning into '/home/jenkins/.jx/environments/****/environment-orioleazure-staging'...
Found remote branch names master, HEAD -> origin/master, master
Switched to branch 'promote-tabletoprank-web-0.0.4'
[promote-tabletoprank-web-0.0.4 3372b38] Promote tabletoprank-web to version 0.0.4
 1 file changed, 8 insertions(+), 5 deletions(-)
fatal: unable to get credential storage lock: Read-only file system
To https://github.com/****/environment-orioleazure-staging.git
 * [new branch]      HEAD -> promote-tabletoprank-web-0.0.4
Created Pull Request: https://github.com/****/environment-orioleazure-staging/pull/1

Using pipeline: ****/tabletoprank-web/master build: #1
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-orioleazure-staging/pull/1 ref 3372b38c983bac4d640b24969c90ace2cc4474ab Could not find a status for repository ****/environment-orioleazure-staging with ref 3372b38c983bac4d640b24969c90ace2cc4474ab
...
```

**Comment by jstrachan** [comment]
@camusd sorry to hear that!

I wonder is there any information in the Pull Request; does the PR CI job complete and be valid? You could always look in Jenkins `jx console` and check out the jobs on the PR of `environment-orioleazure-staging` and on master of `environment-orioleazure-staging` to see if there's any build issues?

**Comment by camusd** [comment]
I figured I'd try creating a cluster on GKE and seeing if I could get a build promoted but I still get an error. Here is the PR CI job build issue I am getting now when I run on GKE:
```
[-dragonglitter-staging_PR-1-EIWUC7VVN4N44PZJGX7JFP25OITEZW77LD6VRZ3OC3NXUQSYCC7A] Running shell script

+ make build

rm -rf requirements.lock

helm version

Client: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}

Server: &version.Version{SemVer:"v2.8.2", GitCommit:"a80231648a1473929271764b920a8e346f6de844", GitTreeState:"clean"}

helm init

Creating /home/jenkins/.helm 

Creating /home/jenkins/.helm/repository 

Creating /home/jenkins/.helm/repository/cache 

Creating /home/jenkins/.helm/repository/local 

Creating /home/jenkins/.helm/plugins 

Creating /home/jenkins/.helm/starters 

Creating /home/jenkins/.helm/cache/archive 

Creating /home/jenkins/.helm/repository/repositories.yaml 

Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 

Adding local repo with URL: http://127.0.0.1:8879/charts 

$HELM_HOME has been configured at /home/jenkins/.helm.

Warning: Tiller is already installed in the cluster.

(Use --client-only to suppress this message, or --upgrade to upgrade Tiller to the current version.)

Happy Helming!

helm repo add releases http://jenkins-x-chartmuseum:8080

"releases" has been added to your repositories

helm repo add jenkins-x http://chartmuseum.build.cd.jenkins-x.io

"jenkins-x" has been added to your repositories

helm dependency build "env"

Hang tight while we grab the latest from your chart repositories...

...Unable to get an update from the "local" chart repository (http://127.0.0.1:8879/charts):

	Get http://127.0.0.1:8879/charts/index.yaml: dial tcp 127.0.0.1:8879: getsockopt: connection refused

...Successfully got an update from the "releases" chart repository

...Successfully got an update from the "stable" chart repository

...Successfully got an update from the "jenkins-x" chart repository

Update Complete. ⎈Happy Helming!⎈

Error: tabletoprank-web chart not found in repo http://jenkins-x-chartmuseum:8080

make: *** [build] Error 1

script returned exit code 2
```

**Comment by camusd** [comment]
I was able to get it to finish the pipeline by modifying the the Jenkinsfile, Chart.yaml, and values.yaml to reference 'tabletoprank-web' instead of 'web'. Not sure why some things were named after the github repo name and some were named after my local git directory.

**Comment by odzhu** [comment]
In my case repository name is correct, however, I get same message

`
WARNING: Failed to query the Pull Request last commit status for To https://github.com/****/environment-pythonlaser-staging.git
 * [new branch]      HEAD -> promote-go2-http-0.0.1
Created Pull Request: https://github.com/****/environment-pythonlaser-staging/pull/3

Using pipeline: ****/go2-http/master build: #1
WARNING: Failed to query the Pull Request last commit status for https://github.com/****/environment-pythonlaser-staging/pull/3 ref 0b85f59181ec016ae6629cfbdb9786142950d602 Could not find a status for repository ****/environment-pythonlaser-staging with ref 0b85f59181ec016ae6629cfbdb9786142950d602
Pull Request https://github.com/****/environment-pythonlaser-staging/pull/3 is merged at sha 12a35480a263fe2783486a0008e2da427607fe2a
Merge commit has not yet any statuses on repo ****/environment-pythonlaser-staging merge sha 12a35480a263fe2783486a0008e2da427607fe2a
`

**Comment by rawlingsj** [comment]
Ok sounds like there's something not quite right with the placholder replacements that happens after the draft create that jx does.  Thanks for reporting, I can take a look at this.

**Comment by jstrachan** [comment]
I think the go placement stuff is fixed now btw

**Comment by eecarus** [comment]
Hi James.  I am also having this problem:

0506d0a7a521545408119b2889f2adff016e5490 Could not find a status for repository *****/environment-mythzircon-staging with ref 0506d0a7a521545408119b2889f2adff016e5490.

I am wondering if there is an update?  I am trying to build on AWS.  In my case, the job just hangs.

**Comment by jstrachan** [comment]
That message is a result of the PR build not completing yet on the environment git repository. I wonder if you look in Jenkins at the `environment-mythzircon-staging` project is the PR build started and still pending? If so it could be your cluster is low on resources maybe? 

If the PR build hasn't even been triggered then its probably webhooks (i.e. github can't trigger Jenkins - can't access it due to firewall issues maybe?). If thats the case you could manually trigger the PR build and when that merges, the pipeline in the `master` branch for `environment-mythzircon-staging`

**Comment by jstrachan** [comment]
@gauravnavgire see my comment above. Could you see if the Pull Request build has been triggered in Jenkins? It could be lack of resources or that webhooks are not working.

**Comment by jstrachan** [comment]
@gauravnavgire @eecarus @odzhu @camusd I've just fixed this issue: #674 so that if, for whatever reason, webhooks are not working/configured/allowed through your firewall, then promotion should still work (albeit slow - it polls every 5 minutes for pending Pull Requests / merges to master)

To try out this fix you can run this:

    jx upgrade cli

which should upgrade your `jx` version to at least `1.2.8` or later to get this fix.

Then either re-create a Jenkins X cluster from scratch or:
* delete your environments via:
```
jx delete env staging
jx delete env production
```
Then remove the jobs from jenkins in the UI
```
jx console
```
Then recreate the Environments again:
```
jx create env staging
....
jx create env production
....
```
then the new projects for the environments should have a background poll of 5 minutes applied. (You can always edit this time to shorter in the Jenkins UI if you `Configure` the environment projects in the Jenkins UI)

**Comment by gauravnavgire** [comment]
@jstrachan Thanks James. Will try the solution you have provided.

And yes I suppose web-hooks are not working.  The pull request check fails as the web-hook is "http://www.unconfigured-jenkins-location.com/job/" instead of "http://jenkins.jx.3x.2xx.xx.xxx.nip.io"

**Comment by eecarus** [comment]
James,

Thank you so much for the quick response.   I will start from a fresh install on AWS and update the CLI like you suggested.  This will insure that I don't have anything else messing me up, like security groups, and the like.

best regards,
Charlie
________________________________

**Comment by eecarus** [comment]
@jstrachan Hi.  Your suggestion worked.   I did bump up the number of nodes on the default Kubernetes cluster from 3 to 5, and patiently waited between each step before proceeding to the next step of the demo.   Following that, I did not have the issue with the hung promote step.   As a second question, is there a way to increase the number of ready Jenkins nodes?  it seems that jobs tend to queue up waiting for another slave pod to be allocated.

BTW, this is GREAT!   My knowledge of Jenkins, fabric8, and  Kubernetes are both so limited that this thing appears like _magic_ to me.  I wish I could do something to help you guys.   

best regards, Charlie

**Comment by jstrachan** [comment]
@eecarus great, glad its working! :) BTW is the delay around a pipeline being triggered in Jenkins and the build pod not starting up; or are you waiting for the job to appear to start inside Jenkins? If its the latter then that sounds like the slow polling thats the backup to webhooks not working. Its best getting webhooks to work so that builds are triggered within a second or so of a git push. 

To diagnose webhooks, if you go into, say, the github repo page then settings then webhooks you can see if github can invoke your jenkins server (there should be webhooks created on each environment or each repo you create/import). It might require whitelisting the github IP addresses to get through your firewall? Another option is to edit the Jenkins projects and reduce the polling from 5 minutes to something much lower - but the worry is you use up your github API limits  if you poll too frequently.

**Comment by jstrachan** [comment]
@eecarus BTW the first build of any kind can be sluggish as the kubernetes nodes need to download the docker images for the build pods etc. You should be able to re-trigger a job directly in the Jenkins UI and it should start executing pretty quick - it can take 10s or so for the pod to actually start running; not sure why its not instant; something to do with the kubernetes-plugin I think

**Comment by eecarus** [comment]
@jstrachan when i recreated the vpc,  i modified the security group to allow 0.0.0.0. so that must have helped!  After you mentioned the webhooks,  I double checked that they existed in each repo and that GH thought it was ok.

I can see the jobs in Jenkins queuing up.   They tend to go pending waiting for an executor.  when an executor pod appears, they’d start running.  When there is a PR promote (so there is 2 merge jobs, one to master on the code, and one to the auto promote environment), it almost looks like both jobs are waiting on each other.  That’s why I was asking if the number of executor pods was controlled by Jenkins or by Kubernetes or by JX.

Cheers!    Thank you again for the fine work.  There are some great ideas here.

________________________________

**Comment by liuxiaopiao** [comment]
I met the error "Error: pipelinetest chart not found in repo http://jenkins-x-chartmuseum:8080" and in monocular it has the chart "tet". 

I am using  jx version 1.2.109.   And the Jenkinsfile, chart.yaml, values are referred with "tet".
Could you please give some comments on this @jstrachan  Thanks a lot.

**Comment by jstrachan** [comment]
@liuxiaopiao that sounds like a separate issue - could you post your pipeline log to a separate issue please?

**Comment by liuxiaopiao** [comment]
Per @jstrachan 's guide, currently we can use the same name for repo and Artifact. The issue is gone. Thanks a lot @jstrachan

**Comment by rohan-sharma-thbs** [comment]
Hi @jstrachan, I had initialized a release build before the PR build was complete (the PR then had stopped in the middle). At the jenkins configuration my pull requests does a clean up and now I do not see older PRs. When I create a new PR that gets built and deployed successfully. How can I create a new release ? Each time it fails when I try to promote with error like:
_**422 Validation Failed [{Resource:PullRequest Field:head Code:invalid Message:}]**_
